engine/player/unique_gear_shadowlands.cpp

```diff
@@ -4280,80 +4273,172 @@ void yasahm_the_riftbreaker( special_effect_t& effect )
+      supremacy_buff->set_can_cancel( false );
+    }
+
+    swap_stat_compensation = dynamic_cast<stat_buff_t*>( buff_t::find( p, "swap_stat_compensation" ) );
+
+    if ( !swap_stat_compensation )
+    {
+      swap_stat_compensation = make_buff<stat_buff_t>( p, "swap_stat_compensation" );
+      swap_stat_compensation->set_duration( 0_s )->set_can_cancel( false );
+    }
 
-  auto buff =
-      make_buff<stat_buff_t>( effect.player, "singularity_supreme", effect.player->find_spell( 368863 ), effect.item );
+    if ( p->antumbra.swap )
+    {
+      swap_stat_compensation->stats.clear();
+      if ( p->antumbra.int_diff != 0.0 )
+        swap_stat_compensation->add_stat( STAT_INTELLECT, p->antumbra.int_diff );
+      if ( p->antumbra.crit_diff != 0.0 )
+        swap_stat_compensation->add_stat( STAT_CRIT_RATING, p->antumbra.crit_diff );
+      if ( p->antumbra.haste_diff != 0.0 )
+        swap_stat_compensation->add_stat( STAT_HASTE_RATING, p->antumbra.haste_diff );
+      if ( p->antumbra.vers_diff != 0.0 )
+        swap_stat_compensation->add_stat( STAT_VERSATILITY_RATING, p->antumbra.vers_diff );
+      if ( p->antumbra.mastery_diff != 0.0 )
+        swap_stat_compensation->add_stat( STAT_MASTERY_RATING, p->antumbra.mastery_diff );
+    }
+
+    set_expire_at_max_stack( true );
+    set_stack_change_callback( [ this ]( buff_t* b, int, int ) {
+      if ( b->at_max_stacks() )
+      {
+        lockout->trigger();
+        supremacy_buff->trigger();
+      }
+    } );
+  }
+};
 
+void singularity_supreme( special_effect_t& effect )
+{
   // despite spell data proc flags, logs seem to show it only procs on damage spell casts
   effect.proc_flags2_ = PF2_CAST_DAMAGE;
-  effect.custom_buff =
-      make_buff<stat_buff_t>( effect.player, "singularity_supreme_counter", effect.player->find_spell( 368845 ), effect.item )
-          ->set_expire_at_max_stack( true )
-          ->set_stack_change_callback( [ lockout, buff ]( buff_t* b, int, int ) {
-            if ( b->at_max_stacks() )
-            {
-              lockout->trigger();
-              buff->trigger();
-            }
-          } );
+
+  auto singularity_buff = new singularity_supreme_t(effect.player, effect.item);
+
+  effect.custom_buff = singularity_buff;
 
   new dbc_proc_callback_t( effect.player, effect );
 
   effect.player->callbacks.register_callback_trigger_function(
       effect.driver()->id(), dbc_proc_callback_t::trigger_fn_type::CONDITION,
-      [ lockout ]( const dbc_proc_callback_t*, action_t*, action_state_t* ) {
-        return !lockout->check();
+      [ singularity_buff ]( const dbc_proc_callback_t*, action_t*, action_state_t* ) {
+        return !singularity_buff->swap_stat_compensation->check() && !singularity_buff->lockout->check();
       } );
 }
 
+// Action to weapon swap Antumbra
+struct antumbra_swap_t : public action_t
+{
+  singularity_supreme_t* singularity_buff;
```

Comment:
What you have here is fine, but personally I think the code would be cleaner if you instead stored the swap_stat_compensation buff here as a buff_t*, and had a stack change callback defined for that buff to handle canceling the other buffs. That would allow all of the code except for this action to be in the `void singularity_supreme( special_effect_t& effect )` block and would not require a dynamic cast in this action.